### DISCUSSION TITLE
Struggles with Region Definitions for Boundary Conditions

### DISCUSSION THREAD

**Discussion by SamKoc15** [lululxvi/deepxde#1538] [General]
Dear Prof. @lululxvi  ,

I have been working on a Inverse Problem for a Pinhole CFD Simulation.
I have created a domain and am trying to define no slip conditions on the walls 1,2,3...10.
(2,4,7,9 are the vertical walls).
Yet when running the example the velocity is unequal to zero at most of these walls as you can see in the image.
Im quite confused why this happens. Has anybody a guess ?

Kind regards and thanks for the help in advance !

![WhatsApp Image 2023-10-30 at 14 31 56](https://github.com/lululxvi/deepxde/assets/149392762/b9b92741-bc07-49dd-bcc0-c322d813b4ba)

```

domain_inlet = dde.geometry.Rectangle(xmin=[0, 0], xmax=[5, 3])
domain_pinhole = dde.geometry.Rectangle(xmin=[5, 1], xmax=[6, 2])
domain_outlet = dde.geometry.Rectangle(xmin=[6, 0], xmax=[21, 3])

inlet_pinhole = dde.geometry.csg.CSGUnion(domain_inlet,domain_pinhole)
spatial_domain = dde.geometry.csg.CSGUnion(inlet_pinhole,domain_outlet)



               1     2      4     5
             |-------|   |-------|
             |          | 3 |          |
             |          |---|         |
   inlet   |                       |  outlet
             |          |---|          |
             |          | 8 |         |
             |-------|   |-------|
               6     7     9     10






def boundary_outlet(x, on_boundary):
    return on_boundary and dde.utils.isclose(x[0], 21)

def boundary_inlet(x, on_boundary):
    return on_boundary and dde.utils.isclose(x[0], 0)

def b_1(x, on_boundary):
    return on_boundary and (np.isclose(x[1], 3) and x[0] < 5)

def b_3(x, on_boundary):
    return on_boundary and (np.isclose(x[1], 2) and x[0] < 6 and x[0] > 5)

def b_5(x, on_boundary):
    return on_boundary and (np.isclose(x[1], 3) and x[0] > 6)

def b_6(x, on_boundary):
    return on_boundary and (np.isclose(x[1], 0) and x[0] < 5)

def b_8(x, on_boundary):
    return on_boundary and (np.isclose(x[1], 1) and x[0] < 6 and x[0] > 5)

def b_10(x, on_boundary):
    return on_boundary and (np.isclose(x[1], 0) and x[0] > 6)

def b_2(x, on_boundary):
    return on_boundary and (x[1] > 2  and np.isclose(x[0], 5))

def b_7(x, on_boundary):
    return on_boundary and (x[1] < 1 and np.isclose(x[0], 5))

def b_4(x, on_boundary):
    return on_boundary and (x[1] > 2  and np.isclose(x[0], 6))

def b_9(x, on_boundary):
    return on_boundary and (x[1] < 1 and np.isclose(x[0], 6))


def b_610(x, on_boundary):
    return on_boundary and (np.isclose(x[1], 0))





bc_inlet_u = dde.icbc.DirichletBC(
    spatial_domain, lambda x: u_inlet, boundary_inlet, component=0)

bc_noSlip_v = dde.icbc.DirichletBC(
    spatial_domain, lambda x: 0. , b_1 and b_2 and b_3 and b_4 and b_6 and b_5 and b_7 and b_8 and b_9 and b_10, component=1)

bc_noSlip_u = dde.icbc.DirichletBC(
    spatial_domain, lambda x: 0., b_1 and b_2 and b_3 and b_4 and b_5 and b_6 and b_7 and b_8 and b_9 and b_10, component=0)

bc_outlet_p = dde.icbc.DirichletBC(
    spatial_domain, lambda x: 0., boundary_outlet, component=2)


data = dde.data.PDE(
    spatial_domain,
    pde,
    [bc_noSlip_v, bc_noSlip_u, observe_u, observe_v,bc_inlet_u],
    num_domain=3000,
    num_boundary=800,
    num_test=100000,
    anchors=xy_train,
```

**Comment by lululxvi** [comment]
Use a larger weight for BC.

**Comment by yuluoxiansen** [comment]
Hello, has your problem been solved? I seem to have encountered the same problem. Mine is the Newman boundary. When the boundary value is 0, the loss value is not good.

**Reply by SamKoc15** [reply]
The problem in this case wasnt the weighting but that the declreation of the boundary conditions just didnt work.
I rewrote it into single lines like 

bc_noSlip_u = dde.icbc.DirichletBC(
    spatial_domain, lambda x: 0., b_1, component=0)
    
bc_noSlip_u = dde.icbc.DirichletBC(
    spatial_domain, lambda x: 0., b_2, component=0)
...
and then used these for the boundary conditions.

If you can upload your code or send it to  I could take look. Are you talking about training or test loss ?

**Reply by yuluoxiansen** [reply]
Sorry for replying to you so late, I'm trying to solve a 2D heat transfer equation with a temperature of 1 on the left side of the area and 0 on the right side, and the rest of the boundaries are adiabatic, I can get a good result when it's a rectangular area, But when I change the area, the loss of Newman boundary does not decrease.

The area is as follows, the parameters are r and rc.
![Snipaste_2023-12-10_17-09-49](https://github.com/lululxvi/deepxde/assets/116873095/f91bb6e5-441c-432d-9811-282da7dd8b69)

My code:

```
import deepxde as dde
dde.config.set_random_seed(1234)
dde.config.real.set_float32()

r = 1
rc = 0.5

vertices = [
    (0, 0), (0, 2 * r), (r, r + rc / 2), (2 * r, 2 * r), (2 * r, 0), (r, r - rc / 2)
]
geoms = dde.geometry.Polygon(vertices=vertices, )

def pde(x, y):
    dy_xx = dde.grad.hessian(y, x, i=0, j=0)
    dy_yy = dde.grad.hessian(y, x, i=1, j=1)
    return dy_xx + dy_yy

def boundary_l(x, on_boundary):
    return on_boundary and dde.utils.isclose(x[0], 0)

def boundary_r(x, on_boundary):
    return on_boundary and dde.utils.isclose(x[0], 2 * r)

def boundary(x, on_boundary):
    return on_boundary and 0 < x[0] < 2 * r

bc_left = dde.icbc.DirichletBC(geoms, lambda x: 1, boundary_l)
bc_right = dde.icbc.DirichletBC(geoms, lambda x: 0, boundary_r)
bc_neumann = dde.icbc.NeumannBC(geoms, lambda x: 0, boundary)

data = dde.data.PDE(
    geoms,
    num_domain=2000,
    num_boundary=200,
    num_test=1000,
    # anchors=X_points,
    pde=pde,
    bcs=[bc_left, bc_right, bc_neumann, ],
)

net = dde.nn.FNN([2] + [50] * 5 + [1], "tanh", "Glorot normal")

model = dde.Model(data, net)
model.compile("adam", lr=0.0005, loss_weights=[1.0, 1.0, 1.0, 1.0], decay=("step", 100000, 0.3,))
losshistory, train_state = model.train(iterations=30000, display_every=1, )

dde.saveplot(losshistory, train_state, issave=True, isplot=True)

```
![Uploading Figure_1.png…]()

Looking forward to your reply!
